types/react/index.d.ts

```diff
@@ -883,7 +883,7 @@ declare namespace React {
     // based on the code in https://github.com/facebook/react/pull/13968
 
     // Unlike the class component setState, the updates are not allowed to be partial
-    type SetStateAction<S> = S | ((prevState: S) => S);
+    type SetStateAction<S> = (S extends (() => any) ? never : S) | ((prevState: S) => S);
```

Comment:
What happens if we pass a function that has parameters?

Reply:
AFAIK, It's OK as any function is assignable to `() => any`

> What happens if we pass a function that has parameters?

Reply:
see [how `react` determine argument is _value_ or _update action_](https://github.com/facebook/react/blob/5f06576f51ece88d846d01abd2ddd575827c6127/packages/react-reconciler/src/ReactFiberHooks.js#L323) ([used in reconciler](https://github.com/facebook/react/blob/5f06576f51ece88d846d01abd2ddd575827c6127/packages/react-reconciler/src/ReactFiberHooks.js#L336))

Reply:
> AFAIK, It's OK as any function is assignable to () => any

This is not accurate:
```tsx
const fn: () => any = (props: object) => {}
```

[Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZmAXDAFAShgXgHwwIZgCeWqADgE4ikTIgBGAVgKbBQY4wDeAvkA)

Reply:
Maybe I confused with other language, and/or `() => any` is assignable to any function. I must have been wrong, thanks for pointing out.

* * *


This change breaks [some test](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/test/next.tsx#L39-L43)

```typescript
    const [func] = React.useState(() => () => 0);

    // $ExpectType () => number
    func;
    // $ExpectType () => number
```

I'll update it ([playground](https://www.typescriptlang.org/play?ts=3.6.3#code/C4TwDgpgBAggdiAYgVzgY2ASwPZygXigAoA6MgNwEMAbZCALikoQEoCA+JhAbgChRIUACKYAzmErA0ACwA8AFU6EiVWgyjy2+TuWyYAJnwHQAchHIQATinRZcCpRqgQAHsAhx9o2AhsYceAD8UHDmVlCM8nz84NAAajR0APKWAKpg+pIQMP72igRQZhbWqLlwDlAAPsREYJbmOMiikVqcmjGCCWopiJQY2JYgFYRFVn525fnVREStGiy8vPoQaNSU9VAAZqUTUKIQwADKwFmyh+wqiRAp6ZnuOROMXclpGVkPAWfsLIy6BnzLVbraDbWwBKBNCDHU7nIiYOCYLA0Z7qFE9PrAAZDc4-KAAbUOABphGIJFI5GjXndsmUvuwALrRAD0TKgACp2ezFiyoAASACiLkgGH5lksA14aFwomA+MoxP2wBg9IKkOh7lksw4IWQAFsAEZWC5a7RQAAMLGiUrgMvx+oVBwAQirCGqThqTZw4HrDZZjXNPebLbweQKhStgKLxZZeIrHTM5gBGABMAGZg5LpbK8WgHcAAMIuiH7dUQTVzb0Go0J7WBi18RX5mumwMp9NWrP4-R5oRFt1ZZucOvBnlRgZQbBoNDISyiEhQQ7SbDIaj6KAASSgguFsrHlkCsYOTcDrbTGZ5HMv3NZAHdpCAJ6uoBJRPtvO4ZUsVmsNqCyo-9C+OEESRagUUYQ4qhqOZDhYXECWJERxEkGRZEpW53lpc4GT4a1bTxCA835ItsFXQdiDmetD0jcjT3bXggA)), but [I'm not sure if I can remove this test](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/f0016483a4b1d1fd6fd636c47eea19e2d1614336#diff-52639bdd9e43a734e9b469af4da781c2e46cf27941a7483555fe5f3978241962R17-R18) now...

Reply:
Breaking existing tests requires some more in-depth analysis. Coudl you provide some real-world scenarios that are improved by your change? How does the change affect existing usage concretely?

Reply:
In the rare case of...

1. using a function as the State **directly**
2. and the type parameter is omitted (rely on inference)
3. and the default value is inferred as a function that returns a literal type

...then new type error will raised on existing code.

* * *

Likewise, in (less) rare cases the advantage: when using a function as the State **directly**, it will raise type error on pass its value (which will identified as a function returns initial value)

```tsx
const initialValue: (() => number) = something;
const [a, setA] = useState<() => number>(initialValue); // type error!
```